File: nvpl_blas/c/ssbmv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_ssbmv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 4;
    nvpl_int_t K = 2;
    float alpha = 1.0f;
    float beta = 1.0f;
    enum CBLAS_ORDER order = CblasColMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    float * A = NULL;
    float * X;
    float * Y;
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 1;

    nvpl_int_t KL = (uplo == CblasUpper) ? 0 : K;
    nvpl_int_t KU = (uplo == CblasUpper) ? K : 0;
    
    nvpl_int_t lda = K + 1;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (N - 1) * labs(incY);

    printf("\nExample: cblas_ssbmv for the symmetric banded matrix-vector multiplication\n\n");
    printf("#### args: n=%" PRId64 ", k=%" PRId64 ", lda=%" PRId64 ", incx=%" PRId64 ", incy=%" PRId64 
            ", uplo=%c, order=%c\n", (int64_t)N, (int64_t)K, (int64_t)lda, (int64_t)incX, (int64_t)incY,
            uplo_to_char(uplo), order_to_char(order));
    printf("           alpha=%g, beta=%g\n", alpha, beta);

    // allocate memory
    A = (float *)malloc(lda * N * sizeof(float));
    X = (float *)malloc(len_x * sizeof(float));
    Y = (float *)malloc(len_y * sizeof(float));

    // fill data
    fill_sbandmatrix(A, KL, KU, N, N, lda, order);
    fill_svector(X, N, incX);
    fill_svector(Y, N, incY);

    // print input data
    print_sbandmatrix(A, KL, KU, N, N, lda, "A", order);
    print_svector(X, N, incX, "X");
    print_svector(Y, N, incY, "Y");

    // call cblas_ssbmv
    cblas_ssbmv(order, uplo, N, K, alpha, A, lda, X, incX, beta, Y, incY);

    // print result
    printf("\nAfter ssbmv operation:\n");
    print_svector(Y, N, incY, "Y");

    // release memory
    free(A);
    free(X);
    free(Y);
    return 0;
}

File: nvpl_blas/f77/ssbmv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     SSBMV 
!
!******************************************************************************
      program   SSBMV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      character*1     uplo
      parameter       (uplo='U')
      integer         n, k, lda, kl, ku
      integer         incx, incy
      parameter       (n=4, k=2, incx=1, incy=1, lda=k+1)
      integer         asize
      parameter       (asize=n*n)
      real            alpha, beta
      parameter       (alpha=1.0, beta=1.0)
      real            a(lda,n)
      integer         xsize, ysize
      parameter       (xsize=1+(n-1)*abs(incx),
     &                 ysize=1+(n-1)*abs(incy))
      real            x(xsize), y(ysize)

*       External Subroutines
      external        print_sbandmatrix, fill_sbandmatrix,
     &                print_svector, fill_svector
*
*      Executable Statements
*
      print*
      print 99, 'Example: SSBMV for the symmetric banded ',
     &          'matrix-vector multiplication'
*
*       Print input data
      if (uplo .eq. 'U') then
          kl=0
          ku=k
      else
          kl=k
          ku=0
      endif
      print*
      print 100, n, k, lda, incx, incy, uplo
      print 101, alpha, beta
      call fill_sbandmatrix(a,kl,ku,n,n,lda)
      call fill_svector(x,n,incx)
      call fill_svector(y,n,incy)
      call print_sbandmatrix(a,kl,ku,n,n,lda,'A')
      call print_svector(x,n,incx,'X')
      call print_svector(y,n,incy,'Y')
*
*      Call SSBMV subroutine
      call SSBMV(uplo,n,k,alpha,a,lda,x,incx,beta,y,incy)
*
      print*
      print 102, 'After ssbmv operation:'
      call print_svector(y,n,incy,'Y')

      stop
 99   format(A,A)
 100  format('#### args: n=',i1,', k=',i1,', lda=',i1,
     &       ', incx=',i1,', incy=',i1,', uplo=',a1)
 101  format(11x,'alpha=',f4.1,'  beta=',f4.1)
 102  format(A)
      end
